src/Network/NWFramer.cs

```diff
@@ -0,0 +1,474 @@
+
+		public NWEndpoint Endpoint => new NWEndpoint (nw_framer_copy_remote_endpoint (GetCheckedHandle ()), owns: true);
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern OS_nw_parameters nw_framer_copy_parameters (OS_nw_framer framer);
+
+		public NWParameters Parameters => new NWParameters (nw_framer_copy_parameters (GetCheckedHandle ()), owns: true);
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern OS_nw_endpoint nw_framer_copy_local_endpoint (OS_nw_framer framer);
+
+		public NWEndpoint LocalEndpoint => new NWEndpoint (nw_framer_copy_local_endpoint (GetCheckedHandle ()), owns: true);
+
+		[DllImport (Constants.NetworkLibrary)]
+		unsafe static extern void nw_framer_async (OS_nw_framer framer, void *async_block);
+
+		delegate void nw_framer_async_t (IntPtr block);
+		static nw_framer_async_t static_ScheduleHandler = TrampolineScheduleHandler;
+
+		[MonoPInvokeCallback (typeof (nw_framer_async_t))]
+		static void TrampolineScheduleHandler (IntPtr block) 
+		{
+			var del = BlockLiteral.GetTarget<Action> (block);
+			if (del != null) {
+				del ();
+			}
+		}
+
+		public void ScheduleAsync (Action handler)
+		{
+			unsafe {
+				if (handler == null) {
+					nw_framer_async (GetCheckedHandle (), null);
+					return;
+				}
+				BlockLiteral block_handler = new BlockLiteral ();
+				BlockLiteral *block_ptr_handler = &block_handler;
+				block_handler.SetupBlockUnsafe (static_ScheduleHandler, handler);
+				try {
+					nw_framer_async (GetCheckedHandle (), (void*) block_ptr_handler);
+				} finally {
+					block_handler.CleanupBlock ();
+				}
+			}
+		}
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern unsafe bool nw_framer_parse_output (OS_nw_framer framer, nuint minimum_incomplete_length, nuint maximum_length, IntPtr temp_buffer, void* parse);
+
+		delegate void nw_framer_parse_output_t (IntPtr block, IntPtr buffer, nuint buffer_length, bool is_complete);
+		static nw_framer_parse_output_t static_ParseOutputHandler = TrampolineParseOutputHandler;
+
+		[MonoPInvokeCallback (typeof (nw_framer_parse_output_t))]
+		static void TrampolineParseOutputHandler (IntPtr block, IntPtr buffer, nuint buffer_length, bool is_complete)
+		{
+			var del = BlockLiteral.GetTarget<Action<byte[], bool>> (block);
+			if (del != null) {
+				var bBuffer = new byte[buffer_length];
+				Marshal.Copy (buffer, bBuffer, 0, (int)buffer_length);
+				del (bBuffer, is_complete);
+			}
+		}
+
+		public bool ParseOutput (nuint minimumIncompleteLength, nuint maximumLength, byte[] tempBuffer, Action<byte[], bool> handler)
+		{
+			unsafe {
+				if (handler == null) {
+					if (tempBuffer == null)
+						return nw_framer_parse_output (GetCheckedHandle (), minimumIncompleteLength, maximumLength, IntPtr.Zero, null);
+					else
+						fixed (void* tmpBufferPtr = tempBuffer)
+							return nw_framer_parse_output (GetCheckedHandle (), minimumIncompleteLength, maximumLength, (IntPtr)tmpBufferPtr, null);
```

Comment:
If you change the P/Invoke to take a byte[] instead of IntPtr, you can simplify a lot of the code here.